### ISSUE TITLE
Insert automatically generated table of contents TOC on rendered markdown files like README.md.

### ISSUE THREAD

**Issue by cirosantilli** [isaacs/github#215]
Edit March 2021: GitHub now added a workaround mentioned at: https://github.com/isaacs/github/issues/215#issuecomment-807688648 That is a good step, but I'll keep this open until they actually add a way to add it inside the rendered output (which can be e.g. searched more easily with Ctrl+F).

When I see a manually generated table of contents, it makes me sad.

When I see a huge README that is impossible to navigate without it, it makes me even sadder.

LaTeX has it. Gollum has it. Pandoc has it. So why not GFM?

There are some tools that automate the generation, but they're just an ugly hack:
- require a Make before pushes
- add output to version control
- modify our dear source code for us

Now the bitter part: what syntax to use?

Whatever is chosen, it should be a standard way to extend Markdown, so that other extensions can be added later on.

One possibility is to use Kramdown extension syntax and insert the TOC with `{:toc max_level=3 }`.

Redcarpet already has a command that generates the TOC: `Redcarpet::Render::HTML_TOC`, but no way to insert it, and no standard way to extend markdown.

Same for GitLab: http://feedback.gitlab.com/forums/176466-general/suggestions/5790538-extension-for-inserting-a-table-of-contents-toc

StackOverflow: 66 upvotes as of 2014-06: http://stackoverflow.com/questions/9721944/automatic-toc-in-github-flavoured-markdown

There is an issue for the library GitHub uses to render markdown at: https://github.com/github/markup/issues/904

[Steven! Ragnarök](https://github.com/nuclearsandwich) replied with the neutral:

> Thanks for the suggestion and links. I'll add it to our internal feature request list for the team to see.

**Asciidoc**

I got tired of waiting and started using Asciidoc. I recommend it, the format is really good. Just look at the typographical beauty of this: https://github.com/cirosantilli/linux-kernel-module-cheat/blob/7d9102373d60bd159920abfe96d636420afedd67/README.adoc

**Comment by szczepiq** [comment]
+1

**Comment by Manuzor** [comment]
+1

**Comment by ErikMHummel** [comment]
+1

**Comment by MikeyBurkman** [comment]
Why does a solution have to extend markdown?

@ekalinin's code can already create a ToC markdown file, so GH could just generate a new ToC on every change to readme.md, and display that file on the project's homepage above the real readme.md. Everyone gets a ToC for free with their readme.md files (and presumably can opt-in/out if they wish) without changing a single line of code.

**Comment by quasipedia** [comment]
> Why does a solution have to extend markdown?

@MikeyBurkman - It does not _need_ to, but if it did it, it would become part of the specification, and all libraries that works against that specification would support it (e.g.: github-markdown preview of the readme file in your IDE/editor of choice).

Preview of the TOC _is_ important: on large documents is easy to get the levels wrong, for example.

**Comment by guillaume-alvarez** [comment]
+1

**Comment by jon-hanson** [comment]
+1

**Comment by eidng8** [comment]
+1

**Comment by MikeyBurkman** [comment]
@quasipedia Surely you aren't saying that the necessity of a preview tool is a good reason to change the MD spec! That's like saying that not having some MD tool installed on my computer means I can't edit MD. Github already have a MD preview tool... just extend it to also preview the TOC.

**Comment by quasipedia** [comment]
@MikeyBurkman - Surely you aren't saying that improving on standards is bad! ;)

Seriously: I can't but repeat myself: the solution does not _need_ to extentd MD.  GitHub has a history of creating standards that are to all effects "forks" of previously existing ones (GFM, TOML...).  Since GFM is already a different standard than MD, and since GitHub is its creator and maintainer, it wouldn't surprise me if they would implement the TOC as part if it...

...but if that doesn't happen, I am just fine! :)

**Comment by ajkochanowicz** [comment]
+1

**Comment by dreamcat4** [comment]
+1

Have tried MarkdownTOC for sublime text today. Which might have been a alternative but sadly it doesn't work nearly well enough. No links and seems broken as it doesn't parse my markdown headings correctly.

**Comment by ajkochanowicz** [comment]
FWIW, I made this. https://github.com/ajkochanowicz/Smooth-ToCer

**Comment by bumbu** [comment]
+1

**Comment by rbs-fueled** [comment]
+1

**Comment by martinheidegger** [comment]
+1

**Comment by sherdeadlock** [comment]
+1

**Comment by wilsonmar** [comment]
+1

**Comment by apjanke** [comment]
+1

**Comment by AndyWendt** [comment]
:+1:

**Comment by v0lkan** [comment]
Bump :+1:

**Comment by idelvall** [comment]
+1

**Comment by dreamcat4** [comment]
Just going to repeat myself here, and say that github should adopt the `doctoc` tool for auto-generating TOCs in markdown files. It's the best one available out there.

**Comment by AndyWendt** [comment]
@dreamcat4 ditto here

**Comment by sc0ttkclark** [comment]
yas

**Comment by jsaz** [comment]
+1

**Comment by np422** [comment]
+1

**Comment by JeremyCaney** [comment]
+1

**Comment by dscho** [comment]
My colleague @schmiedc and I have created [a GreaseMonkey script](http://tomancaklab.github.io/gfm-add-toc.user.js) that installs a new `TOC` button left of the `h1` button which uses the excellent `markdown-js` library to add/refresh a table of contents.

The advantage over solutions like doctoc is that it integrates into GitHub's wiki editor and does not need users to work on their command-line (and require users to install tools like `node.js`). In Chrome, it works by drag 'n dropping into the Extensions page, in Firefox you will need to install the GreaseMonkey extension.

It will work with plain markdown (i.e. it does not handle code blocks correctly, as that is a GitHub extension to markdown). Contributions welcome.

**Comment by bumbu** [comment]
@dscho does it work only with wiki pages or with any markdown file?

**Comment by dscho** [comment]
@bumbu we specifically designed it to trigger only when the user edits wiki pages. But it should be really trivial to adjust by appending appropriate [`@include` lines](https://github.com/tomancaklab/tomancaklab.github.io/blob/ae6342811589a0c92168fbc8d2606a72c2992eea/gfm-add-toc.user.js#L4-L5) and extending [this regular expression](https://github.com/tomancaklab/tomancaklab.github.io/blob/ae6342811589a0c92168fbc8d2606a72c2992eea/gfm-add-toc.user.js#L11) to accept more URLs than the "edit wiki page" ones.

Why not give it a shot yourself? As I mentioned, contributions are warmly welcomed.

**Comment by dmitrizagidulin** [comment]
+1

**Comment by S0lll0s** [comment]
+1

**Comment by abalmos** [comment]
+1

**Comment by pepaslabs** [comment]
+1

**Comment by ekalinin** [comment]
>> @ekalinin I suppose this won't work for private repos?
>
> @mike-charles yes, unfortunately [github-markdown-toc](https://github.com/ekalinin/github-markdown-toc) won't work for private repos
>

@mike-charles just added an ability to work with local markdown files. So seems it may work with private repos.

**Comment by mike-charles** [comment]
@ekalinin thanks!

**Comment by 4F2E4A2E** [comment]
+1

**Comment by AzizSaleh** [comment]
+1

**Comment by nyh** [comment]
+1

**Comment by AshCoolman** [comment]
+1

**Comment by art-divin** [comment]
+1

**Comment by OpenESignForms** [comment]
+1 for an auto-generated TOC.  All of my wikis ported from google code are much less useful with the TOC, but manually building them is a huge effort and will be error prone to say the least.

**Comment by lebolo** [comment]
:+1:

**Comment by cederom** [comment]
still no [TOC] ? even sourceforce has it! :-) +1

**Comment by mguida22** [comment]
:+1: this would be very helpful

**Comment by dscho** [comment]
OMG if the GitHub people are as annoyed as I am by all those +1 mails spamming the inbox, this feature is never going to happen... :grinning:

**Comment by dreamcat4** [comment]
On Mon, Apr 13, 2015 at 10:13 AM, dscho <notifications@github.com> wrote:

> OMG if the GitHub people are as annoyed as I am by all those +1 mails
> spamming the inbox, this feature is never going to happen... [image:
> :grinning:]
>
+1

**Comment by dscho** [comment]
@dreamcat4 LOL!

**Comment by p8952** [comment]
+1

**Comment by AndyWendt** [comment]
@dscho and @dreamcat4 here's another :+1:  :smile: that made me laugh

**Comment by robbyoconnor** [comment]
:+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1:

**Comment by ogradyjd** [comment]
I just ran into this and behold, a feature request with never-ending +1s.  Are the devs trying to tell us that ASCIIdoc is the way to go?

Also: 

+!

**Comment by dscho** [comment]
> I just ran into this and behold, a feature request with never-ending +1s. Are the devs trying to tell us that ASCIIdoc is the way to go?

Yeah, those +1s really hide the important comments, don't they? In particular, you could work around it locally by [installing GreaseMonkey script, as I described above](https://github.com/isaacs/github/issues/215#issuecomment-78488786) -- this works in Chrome and in Firefox, even on Android.

> Also:
> 
> +!

Hahaha! You caps-locked the "1"... :grinning:

**Comment by AndyWendt** [comment]
+! > +1 :smile:

**Comment by codevinsky** [comment]
Excuse me, but can I sit on your keyboard? :cat2: 

+1

**Comment by eterps** [comment]
:+1:

**Comment by JamesLavin** [comment]
Everyone's sick of all the +1s on this thread, so I'm issuing a joint +2 with my colleague Pinny.

**Comment by Franck-Dernoncourt** [comment]
:+1:

**Comment by kdemarest** [comment]
+1

**Comment by einverne** [comment]
+1

**Comment by robbyoconnor** [comment]
FYI you can use emacs to generate a toc but something like this in the 
Web UI would be nice.

**Comment by svoop** [comment]
+1

**Comment by bkuberek** [comment]
:+1:

**Comment by theofidry** [comment]
+1

**Comment by jayschwa** [comment]
+1

**Comment by breakingco** [comment]
:( :+1

**Comment by nelson6e65** [comment]
:+1:

**Comment by ikitommi** [comment]
:+1:

**Comment by ryanhoo** [comment]
If you want to make a sense of the structure of whole article, you really need a nice fixed table of contents on either the left/right or top of the page.

**Comment by davelnewton** [comment]
:+1: although in theory I should be rolled into JamesLavin's +2, making it a +3, and eliminating the need for this :+1:.

**Comment by shadowbq** [comment]
+1 please.

**Comment by alejdg** [comment]
+1

**Comment by archiecobbs** [comment]
+1

**Comment by Delawen** [comment]
Probably, after a year ignoring this, it won't be done. But :+1:

**Comment by adamwdennis** [comment]
+1

**Comment by sebask** [comment]
+1

**Comment by amaiorano** [comment]
I wrote a python script to generate a TOC from a markdown file: https://github.com/amaiorano/md-to-toc

This will work on any markdown file - doesn't need to be on a public GitHub repo.

**Comment by josephjoaquinmartin** [comment]
@sebask had a lot of good things to say. I will just add my own two cents:

+1

**Comment by toshimaru** [comment]
++1

**Comment by likejazz** [comment]
+1

**Comment by beckybeq** [comment]
+1

**Comment by cornelone** [comment]
+1

**Comment by cyChop** [comment]
+1

**Comment by AndyWendt** [comment]
:+1:

**Comment by ricardoamaro** [comment]
+1

**Comment by posix4e** [comment]
+1!

**Comment by kdemarest** [comment]
I looked into creating a web service that would read the .md file, generate the TOC and return appropriate insertable content, but of course the markdown is too locked down for that.

However, I did find the location of the code. Perhaps someone will beat me to the punch creating a PR for this:

```
https://github.com/github/markup
```

**Comment by RuedigerMoeller** [comment]
+5

**Comment by AndyWendt** [comment]
+99

**Comment by rektide** [comment]
https://github.com/isaacs/github/issues/215

**Comment by leissa** [comment]
My current workaround:
```sh
#!/bin/sh
#
# A pre-commit hook to automatically generate a table of contents on top of each *.md file.
# you need doctoc for this hook to work:
# https://github.com/thlorenz/doctoc

doctoc --title "**Contents**" --github .
```
put this in ```.git/hooks/pre-commit```

see, e.g., here: https://github.com/AnyDSL/anydsl/wiki/Programming-Guide

**Comment by maxless** [comment]
Well, a year has passed...

**Comment by Kobold** [comment]
+1

**Comment by benhourigan** [comment]
+1

**Comment by shreyansb** [comment]
+1!!!

**Comment by AndyWendt** [comment]
:+1: +1 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by ObjectiveTruth** [comment]
:+1:

**Comment by saurabh-agrawal83** [comment]
+1

**Comment by atesgoral** [comment]
:+1:

**Comment by leafpeak** [comment]
ETA????!!!

**Comment by blakehawkins** [comment]
:+1:

**Comment by aldanor** [comment]
!

**Comment by jwaldrip** [comment]
+1

**Comment by TWiStErRob** [comment]
+1

**Comment by reckart** [comment]
+1

**Comment by brunobowden** [comment]
Can we have a more meaningful update on this bug? Ideally a statement from GitHub about whether they'll be working on this and if so, then over what time frame.

I think it would be better to lock down the bug as the numerous +1's are noisy rather than helpful. You can still lend your weight in support by subscribing to the bug. The other comments with more detail have been helpful.

**Comment by shadowbq** [comment]
>>  Hi shadowbq,
>>
>> Thanks for writing in, and sorry for the frustration. As the user 'isaacs' states in his repo's readme,
>> the issues listed there are not official GitHub issues or WIP items. We have had several users 
>> request this feature, and we'll add your +1 to this request, but we can't say if/when the feature will 
>> be implemented since our product roadmap is determined by a large number of factors. The team 
>> will definitely read through your concerns though, so thanks again for your email.
>>
>> Cheers,
>> The GitHub Support Team

**Comment by ourmaninamsterdam** [comment]
:+1: :+1:

**Comment by Gawen** [comment]
:+1:

**Comment by tolgaek** [comment]
:+1:

**Comment by PiotrJustyna** [comment]
:+1:

**Comment by 304NotModified** [comment]
+1

**Comment by florianajir** [comment]
+1

**Comment by hohwille** [comment]
It used to work fine with asciidoc using
```
:toc:
toc::[]
```

However this stopped to work recently. :-1:

**Comment by pando85** [comment]
+1

**Comment by brunobowden** [comment]
You can generate a TOC for a markdown page, by opening the javascript console on the public page and then running the following script. It will output the links that you can then enter in to your document. You may need to adjust the 'article>h3' setting depending on your hierarchy. It would of course be much better if this was generated automatically or before commits.

$('article>h3').map(function(i){
    var txt = $(this).text();
    var href= $(this).find('a').attr('href');
    return "- [" + txt + "](" + href + ")";
  }).get().join('\n');

**Comment by 304NotModified** [comment]
@brunobowden good idea. 

It's not fully working. Maybe post it with [fenced code blocks](https://help.github.com/articles/github-flavored-markdown/#fenced-code-blocks)?

**Comment by brunobowden** [comment]
Thansks @304NotModified - I should know better to do that. This works on a public page but it would be helpful to adapt it to run on preview pages as well, so you can do the update from a PR before it's merged.

```javascript
  $('article>h3').map(function(i){
    var txt = $(this).text();
    var href= $(this).find('a').attr('href');
    return "- [" + txt + "](" + href + ")";
  }).get().join('\n');
```

**Comment by aubergine10** [comment]
> I prefer pages that are impossible to navigate, especially when trying to learn how to use a new code library, it makes finding information so much harder and delightfully tedious. /sarcasm

It's somewhat bemusing that GFM can't even muster a table of contents. Makes me wonder if markdown is the right tool for the job?

**Comment by stereobooster** [comment]
https://github.com/gollum/gollum/wiki#table-of-contents-toc-tag

**Comment by brunobowden** [comment]
Gollum only works on the Wiki and not on the markdown files in the main
repo.

**Comment by flyhigher139** [comment]
+1

**Comment by GrahamTheCoder** [comment]
:+1:

**Comment by aioue** [comment]
+1

**Comment by altsheets** [comment]
[[toc]] please

**Comment by bbozo** [comment]
[[toc]] for world peace :+1:

**Comment by binarytemple** [comment]
+1 cmon guys... bitbucket have had this since the dawn of time

**Comment by perryism** [comment]
:+1:

**Comment by 304NotModified** [comment]
PS

>Post an issue if you have an issue or feature request for GitHub. But you should also email support@github.com, since this repo is strictly for our own (unofficial) tracking purposes

**Comment by stresslimit** [comment]
+1 we've moved to `.adoc` to get this, but it's annoying compared to a lot of GFMD features.

**Comment by wbharding** [comment]
140 developers and counting, github. +1
Doesn't seem like rocket science here.

**Comment by altsheets** [comment]
On 17 September 2015 at 06:10, GitHub Staff <support@github.com> wrote:
> [[toc]] please

Thanks for the suggestion, we'll definitely pass that on to the team to consider. 
Let us know if there's anything else we can help you with!

Cheers,
GitHub Support

**Comment by mickep76** [comment]
+1

**Comment by zfedoran** [comment]
+1

**Comment by Paulche** [comment]
+1

**Comment by zackhsi** [comment]
+1

**Comment by Sebb767** [comment]
+1

**Comment by nebezb-ssense** [comment]
+1

**Comment by agrbin** [comment]
+1

**Comment by btiernay** [comment]
+1

**Comment by fwmechanic** [comment]
+1

**Comment by funkhou9** [comment]
[[toc]] :+1:

**Comment by makebbekus** [comment]
+1

**Comment by lklots** [comment]
+1

**Comment by yannickwurm** [comment]
+1

**Comment by clajder** [comment]
+1

**Comment by natsn** [comment]
[[toc]] :+1: pleeeaase

**Comment by DanielSantoro** [comment]
Fully support this - I write _every_ WordPress post, business document, readme, and really anything with text with Markdown and this is a huge sore spot. Having to use an external tool (or even a bigger PITA, coding by hand) takes a lot away from what makes Markdown so great.

Let's see it, GitHub! :)

**Comment by conan** [comment]
+1

**Comment by frhe** [comment]
+1

**Comment by apsey** [comment]
+1

**Comment by cdman** [comment]
+1

**Comment by inversion** [comment]
+1

**Comment by kungfujam** [comment]
+1

**Comment by tobyreid** [comment]
+1

**Comment by alex-netquity** [comment]
+1

**Comment by jansepar** [comment]
:+1:

**Comment by haslo** [comment]
:+1:

**Comment by whyleee** [comment]
+1

**Comment by alexkleinubing** [comment]
+1

**Comment by v0lkan** [comment]
::thumbsup::

**Comment by v0lkan** [comment]
::thumbsup::

**Comment by v0lkan** [comment]
:+1:

**Comment by aguynamedben** [comment]
:+1:

**Comment by jtechau** [comment]
+1

**Comment by csaftoiu** [comment]
+1

**Comment by okbonner** [comment]
+1

**Comment by sholsapp** [comment]
:pray:

**Comment by allenmanning** [comment]
+1

**Comment by melrief** [comment]
+1

**Comment by tpkelly** [comment]
+1

**Comment by keremgocen** [comment]
:beer:

**Comment by dobrescu** [comment]
+1

**Comment by nahidakbar** [comment]
+1

**Comment by richbrat** [comment]
+1

**Comment by mhulse** [comment]
Um, +1.

**Comment by cniedzwiedz** [comment]
+1

**Comment by mckaygerhard** [comment]
there's some well yet implemented solutions to auto generated contents for git based wikis, but currently gitlab to make happy guindon-related They are taking too much time to implementing...

what happened @suport ? if this implementacion If this were open source, the community long ago he had completed!!!

**Comment by lankz** [comment]
+1

**Comment by PullJosh** [comment]
+1

**Comment by famzah** [comment]
+1

**Comment by kristovatlas** [comment]
+1

**Comment by brancomat** [comment]
+1

**Comment by Rusik** [comment]
+1

**Comment by jo32** [comment]
+1

**Comment by jirik** [comment]
+1

**Comment by yanli0303** [comment]
:+1:

**Comment by marafa** [comment]
one and a half years and no solution? 
and yes i am subscribing with that comment

**Comment by xca1019** [comment]
+1

**Comment by takuya** [comment]
+1

**Comment by andreagoulet** [comment]
+1

**Comment by filipsPL** [comment]
+1

**Comment by nickiemc** [comment]
+1

**Comment by joejohndoe** [comment]
+1

**Comment by frogsapo** [comment]
+1

**Comment by MikkelPaulson** [comment]
+1

**Comment by kian** [comment]
+1

**Comment by sergeyklay** [comment]
+1

**Comment by flozzone** [comment]
gollum/gollum-lib#163
gollum/gollum-lib#182

**Comment by chenson42** [comment]
+1

**Comment by siddtewari** [comment]
:+1:

**Comment by Frosty-Z** [comment]
:+1: +1

**Comment by sc0ttkclark** [comment]
Nearly 180 people have +1'd this, I'm not even including those who have already unfollowed the issue due to constant +1's :)

    Post an issue if you have an issue or feature request for GitHub. But you should also email support@github.com, since this repo is strictly for our own (unofficial) tracking purposes

Be sure to send an e-mail out to them as well!

**Comment by dtmilano** [comment]
I obtained decent results using **MediaWiki** edit mode and using the `__TOC__` tag.

**Comment by mhulse** [comment]
> I obtained decent results using MediaWiki edit mode and using the `__TOC__` tag.

Interesting. That didn't seem to work via a wiki page.

Does that work on a README file?

**Comment by dtmilano** [comment]
Yes, it works on a wiki page.

This MediaWiki markup:

```
__TOC__

== GET method ==
=== General ===
==== culebra ====
Invokes the main culebra UI.
...
```
gives something like this:
![selection_056](https://cloud.githubusercontent.com/assets/982482/11634401/3ca10268-9cde-11e5-864d-29aecbe71c29.png)

**Comment by mhulse** [comment]
> Yes, it works on a wiki page.

Ah, then maybe what I want isn't possible. I'd like an auto TOC to generate, on the homepage of my wiki, that lists all pages in wiki (not just the sections on same page).

**Comment by onikienko** [comment]
:+1:

**Comment by sergiopereiraTT** [comment]
Please GitHub! You can do this! :octocat:

**Comment by pszabop** [comment]
YAPO  (Yet another +1)

**Comment by at-huydq** [comment]
Come on Githubbbb +1

**Comment by humbertosanchezlara** [comment]
+1

**Comment by Ky6uk** [comment]
:+1:

**Comment by rtaph** [comment]
+1

**Comment by MandaBrown** [comment]
:+1:

**Comment by oldergod** [comment]
+1

**Comment by johnyf** [comment]
:+1:

**Comment by dmehra** [comment]
+1

**Comment by tonypujals** [comment]
+1 for page level TOC.

**Comment by calvinxiao** [comment]
+1

**Comment by yas375** [comment]
+1

**Comment by gitawego** [comment]
+1 like in bitbucket

**Comment by atodorov** [comment]
+1

There's lots of +1s here. How do we get GitHub to take action on this?

**Comment by msokalski** [comment]
+TOC

**Comment by hoffm** [comment]
+1

**Comment by herikwebb** [comment]
+1

**Comment by yas375** [comment]
While we all are waiting on this, I'd like to share what I've started to use to solve this: 

https://github.com/thlorenz/doctoc
After editing a README I only have to run `doctoc .` and it regenerates the TOC.

**Comment by calvinxiao** [comment]
Guys there is Google Chrome extention that can solve this problem for me. [https://chrome.google.com/webstore/detail/github-toc/nalkpgbfaadkpckoadhlkihofnbhfhek](https://chrome.google.com/webstore/detail/github-toc/nalkpgbfaadkpckoadhlkihofnbhfhek)

**Comment by IEvangelist** [comment]
:+1:

**Comment by dgrabla** [comment]
+1

**Comment by codingberlin** [comment]
+1

**Comment by MikeRatcliffe** [comment]
+1... come on guys, it would only take a few minutes and is clearly a much requested feature.

**Comment by mhulse** [comment]
> +1... come on guys, it would only take a few minutes and is clearly a much requested feature.

Anyone know if GitHub peeps read the comments in this repo?

I wonder how many of us have actually e-mailed to bug them about this?

From this repo's readme:

> Post an issue if you have an issue or feature request for GitHub. **But you should also email support@github.com, since this repo is strictly for our own (unofficial) tracking purposes.**

I'd hate to bombard GitHub with the same request, but do they take quicker action if they get a ton of requests via e-mail? Or, is there another way to help them make the decision to expidite an issue? :dancers: 

:octocat:

**Comment by AzizSaleh** [comment]
Or just forward this thread via email.

**Comment by morotti** [comment]
+1

**Comment by fommil** [comment]
:+1:

**Comment by 1st** [comment]
It will be goof to have ability to insert something like this at the top of README.md file to generate TOC based on headers in this document.

```
Table of Content
-------

[[TOC]]
```

This instruction should handle ALL HEADERS that come AFTER this `[[TOC]]` instruction to avoid including headers that appear before this TOC.

Any plans to implement this simple improvement to GitHub README.md markdown?

**Comment by arnaugm** [comment]
+1

**Comment by splittingred** [comment]
+1

**Comment by gergelyorosz** [comment]
+1

**Comment by mindw** [comment]
+1

**Comment by ninrod** [comment]
this would be highly appreciated.

**Comment by dneise** [comment]
+1

**Comment by jasonwr** [comment]
+1 Send an email to: support@github.com and link in this thread.

**Comment by abalter** [comment]
+1

**Comment by ealves-pt** [comment]
+1

**Comment by openwis-ss** [comment]
Come on, TOC +111

**Comment by markholdt** [comment]
+1

**Comment by JabbyPanda** [comment]
TOC is helping to deal with a long documentation pages :100:

**Comment by hashemian** [comment]
+1 that seems to be something quite straightforward to do and with so many people in need for it. I hope we will see this coming soon.

**Comment by gbirke** [comment]
+1

**Comment by Owlz** [comment]
+1

**Comment by red2678** [comment]
+1

**Comment by hugocf** [comment]
:+1:

**Comment by tmartin1** [comment]
+1

**Comment by TylerRick** [comment]
:+1:

**Comment by zoubeiri** [comment]
+1

**Comment by svoop** [comment]
#### Please don't +1 anymore. You can leave your like by clicking on the +:smile: icon on the top comment or – even better – let Github know by sending them an email, thanks!

**Comment by AzizSaleh** [comment]
I don't think github cares enough about this feature imo, this thread been
going on for months (and that was only after I replied/subscribed to it).

Maybe this would have weight on it if more users with paid plans sent those
emails otherwise it wouldn't go on being ignore for going almost for 2
years.

**Comment by gollogly** [comment]
I have sent them a direct request via https://github.com/contact and will update anything I find out. Maybe this is not the right channel and this is not the right way to go about it.

**Comment by gollogly** [comment]
Hi there, here is the reply and quite prompt which was nice.


Hello Martin,

Thanks for your feedback about the wiki TOS. I personally think it would be a useful addition and I would use it myself.

Currently we aren't able to comment on the status of feature requests, but I will pass on your support for this issue.

Regards,
GitHub Support

**Comment by IEvangelist** [comment]
I did the same thing, email them directly awhile ago -- here is the response I received:

> Thanks for expressing interest in this, David -- I'll add another +1 for this internally.
> 
> Cheers,
> Ivan

Apparently they have their own internal +1 system :angry:

**Comment by IEvangelist** [comment]
I have found [**this**](https://chrome.google.com/webstore/detail/github-toc/nalkpgbfaadkpckoadhlkihofnbhfhek) to be very useful though, it's only limitation is that it requires [**Google Chrome**](https://www.google.com/chrome/browser/desktop/index.html).

**Comment by ebekebe** [comment]
+1

**Comment by agrosner** [comment]
almost 2 years later....all you need to do is parse the ### titles and organize them from single # to lowest hash under a higher value.

**Comment by james-c-barnes** [comment]
+1

**Comment by sidja** [comment]
+1

**Comment by Tuurlijk** [comment]
+1

**Comment by ajpevers** [comment]
Also mailed to support@github.com

**Comment by ajpevers** [comment]
Hey Anton,

Thanks for reaching out. Yes, that is a very popular request. Thanks for letting us know you are interested in it.

I can't make any promises, but I will share your support with the team for consideration.

Regards,
GitHub Support

**Comment by kainr2** [comment]
Yep, after two years, it would be great to finally see TOC.  It is a life-saver when I have a technical FAQ page, and technical notes page.

**Comment by timaschew** [comment]
I created a small, minimal tool which generates a TOC at runtime, of course this won't work on the github.com domain, only on your own (sub)domains like gh-pages. If you use Jekyll there is no need for it, because of Kramdown, but if you don't use Jekyll or Kramdown it can help. 
I listed other tools in my readme as well.

http://timaschew.github.io/auto-toc.js/

**Comment by Mottie** [comment]
I just created a userscript that adds a Table of Contents to readme & wiki pages. 

https://github.com/Mottie/GitHub-userscripts/wiki/GitHub-table-of-contents

This screenshot is of the script working on the [Sinatra repo main page](](https://github.com/sinatra/sinatra))

![github-toc](https://cloud.githubusercontent.com/assets/136959/14096460/8f05ae3e-f52e-11e5-9194-8506ffb2068e.gif)

If you need help with installation instructions, see [this readme](https://github.com/Mottie/GitHub-userscripts#installation).

**Comment by smougenot** [comment]
+1

**Comment by mweitzman** [comment]
+1

**Comment by bencaradocdavies** [comment]
+1

Foswiki and TWiki have had %TOC% for many years. MediaWiki goes further and automatically adds a table of contents to any page with more than three sections. A table of contents is an absolutely essential feature to support wiki pages that are longer than a screen or two.

**Comment by qiuxiafei** [comment]
+1

**Comment by pkrauss-asideas** [comment]
+1

**Comment by chaseWillden** [comment]
+1

**Comment by ldeviveiros** [comment]
+1

**Comment by nishantshobhit** [comment]
+1

**Comment by mautrix** [comment]
+1

**Comment by rylectro** [comment]
+1

**Comment by aoyama-val** [comment]
+1

**Comment by aensley** [comment]
Something like the [Table of Contents for GitHub Chrome Extension](https://chrome.google.com/webstore/detail/table-of-contents-for-git/hlkhpeomjgelmljaknhoboeohhgmmgcn) is what I would expect GitHub to have natively:

An automatically-generated, hierarchical table of contents based on document headers. It should be available for any document type with headers (Markdown, HTML, Readme, Wiki, etc.).

**Comment by EugenMayer** [comment]
Really missing this feature to make the readme even more valuable to those, needed them most. Many issues would not even be created if people could have the oversight and thus read the sections they really need

**Comment by lorderikir** [comment]
+1

**Comment by daortizh** [comment]
Please add suppport for [TOC]

**Comment by Mottie** [comment]
Alternatively, I just created a [userscript that collapses headers within markdown](https://github.com/Mottie/GitHub-userscripts/wiki/GitHub-collapse-markdown). It works on readme, issues, markdown preview, wiki & gist pages.

![github-collapse-markdown](https://cloud.githubusercontent.com/assets/136959/16364748/34f5d776-3bb4-11e6-84c7-a00f2817f80a.gif)

Screenshot from [this page](https://github.com/sinatra/sinatra).

**Comment by tukkek** [comment]
Just letting you all know there was a great issue request written in an official github repository (`github/markup`) for this TOC problem.

If we could get all the users who left comments here to go and "like" the opening comment there I'm certain this would be very helpful - just please lets try not to flood the issue with "+1" comments since we've gone enough of that here already. If you feel you must then post a meaningful comment that adds to the discussion and link back to this thread here as I've done just a moment ago.

https://github.com/github/markup/issues/904

**Comment by TPS** [comment]
@tukkek Extraordinary find! 🙇

@isaacs @cirosantilli Based on https://github.com/github/markup/issues/904#issuecomment-231372455, this issue should get the <kbd>__Parity__</kbd> label, but based on @tukkek's [comment just above](#issuecomment-231374124), it also probably should be closed (& perhaps locked) in favor of that **_official_** issue, to have folks work with @GitHub directly there.  What do you all think?  Is there an official policy &/or label for such a case?

**Comment by cirosantilli** [comment]
@TPS why parity?

I think we should keep this open, as it is not clear that the issue is in that library itself / yet another library like redcarpet + it is good to keep those upvotes open to give more weight to this issue :-)

**Comment by cirosantilli** [comment]
@TPS this issue is specific about README, I think parity should only be added if a competitor does RAEDME ToCs.

**Comment by TPS** [comment]
If you're _that_ strict, you're correct. 😜🙇

**Comment by tukkek** [comment]
> so, inductively, @GitHub's competitors (in the wiki arena) have this option, while @GitHub itself doesn't (but should)!

Actually GH has it but just not on redcarpet's markdown, which is their default markup option.

> this issue is specific about README

Technically, yes but if you read through the comments you'll see that is not really the case. I for one am only interested in this for the wiki feature on GH. I propose this issue should be rename instead to something like "enable TOC on markdown" (in which case it could be closed in favor of the official one).

Nonetheless I'm not positive on it but I think the rendering/transformation tool would work both on wiki and MD files (README) once implemented. I can't imagine there being very different operations for this sort of thing.

Also I'm planning to send an e-mail to GH staff and demand a position since that is their official repository for markup and it's shameful they haven't reacted to the feedback at all for weeks. But I don't want to get ignored (or mostly ignored) so I'm waiting for more likes to pop-up there - so if you're reading this and haven't liked the opening comment there please take a minute of your time to do so: https://github.com/github/markup/issues/904

I'm monitoring the issue there. As soon as the likes stop coming for a day or two I'll write to them. I was hoping we'd have 100 likes so they'd really be forced to make a position. I thought it would be easy with 251 participants here but with 20 likes so far it doesn't sound it'll go much higher than that.

**Comment by asbjornu** [comment]
:thumbsup:

**Comment by ForbodingAngel** [comment]
Wow, over 2 years. For fucks sake.

**Comment by andychups** [comment]
So?

**Comment by kenorb** [comment]
This issue should be closed in favor of github/markup/issues/904

**Comment by wilsonmar** [comment]
@kenorb this issue should be worked on. Closing this would make it seem like it's a new issue when in fact it's been several freaking YEARS with many +1's. Is this really so hard?

**Comment by kenorb** [comment]
@wilsonmar The difference is that this issue is meaningless and nobody from GitHub can look at it. It only generates more noise, since everything is going to /dev/null. Reporting this at github/markup repo, it can be looked at by GitHub staff and the official decision can be made, here it won't be possible and it only confuse people.

**Comment by AzizSaleh** [comment]
@kenorb They know, they just won't give it a medium priority. More than few people emailed support getting only canned responses. I think the fact that majority of the users who want it are not premium has something to say about it. It makes sense to fill in requests from paid users if you want to keep them.

**Comment by kenorb** [comment]
@AzizSaleh I've two premium accounts, but I don't think I've any priority in terms of any requests in comparison with others.

**Comment by Lemongrass3110** [comment]
+1

**Comment by franslundberg** [comment]
+1 from me too. This is really important to us. We (ASSA ABLOY, the largest lock manufacturer in the world) do use paid Github accounts with Markdown documentation. It works well except for the lack of automatic TOC generation. This is essential, please implement it.

**Comment by TPS** [comment]
@franslundberg You should mention all that at the official issue @ https://github.com/github/markup/issues/904, where it'd have more clout.  This repo, of course, is just self-help & griping (& officially to show support, too :wink:).

**Comment by cryptographrix** [comment]
+1

**Comment by shouh** [comment]
To generate TOC with a tool is a hassle. I think that GitHub should support this automatic generation bacause easy to using TOC. 

For example.

(Idea1): With GFM, write `[:contents]` like this.

```
# Readme

[:contents]

## Demo
...

## Requirement
...
```

(Idea2): With a repo's setting like this.

0. Open https://github.com/USERNAME/REPONAME/settings
0. Check `Features > TOC Automatic Generation` from OFF to ON.

**Comment by Mithgol** [comment]
+1

(This comment is a tool that brings this issue from https://github.com/notifications to https://github.com/notifications/participating for me. Just raising my personal awareness.)

**Comment by nepaul** [comment]
+1

**Comment by andris** [comment]
+1

**Comment by TPS** [comment]
@jlengstorf [has just commented @ the official issue that this is _already_ implemented.](https://github.com/github/markup/issues/904#issuecomment-267380899)…  Anyone confirm, please?

**Comment by NobbZ** [comment]
He has not. He only mentioned it is available in jekyll and proposed
similar syntax for repositories read me files.

TPS <notifications@github.com> schrieb am Do., 15. Dez. 2016, 19:29:

**Comment by cpreid** [comment]
+1

**Comment by DongShi** [comment]
+1 
An important feature, it is nice to have it supported.

**Comment by ckpinguin** [comment]
+1

**Comment by William-Sang** [comment]
+1

**Comment by vphantom** [comment]
GitHub should set up some kind of anti "+1" comment filter now that proper reactions are built-in.

**Comment by Mithgol** [comment]
@vphantom “+1” comment is not “reaction only”: it also brings the issue to the page https://github.com/notifications/participating (and a comment is the only way to do that; notice that the “Subscribe” button brings the page only to https://github.com/notifications and no further).

**Comment by vphantom** [comment]
Ah that's a shame, because they sure do pollute popular issues like this one along with their followers' mailboxes.  I had to unsubscribe from a few, short of being able to filter out the noise.  Anyway, sorry for the aside; I didn't mean to derail from topic…

**Comment by AzizSaleh** [comment]
In 4 months this request would hit 3 year mark being ignored. I am pretty sure that Github doesn't think this issue is important and I doubt it will even be done. I will check back in a couple of years.

You can also +1/heart the issue: https://github.com/github/markup/issues/904 maybe that will give it some traction.

**Comment by jhabdas** [comment]
This is a userland problem IMO and should be solved using a static site generator. Consider enabling GitHub Pages on your sites and leveraging a [Jekyll plug-in](https://jekyllrb.com/docs/plugins/) to generate your TOC and improve your documentation. @cirosantilli would you please close this issue?

**Comment by ForbodingAngel** [comment]
@jhabdas Bruh, do you even wiki? Literally every other wiki on the planet has the ability to autogenerate TOC dynamically. Your suggestion and assertion is pants on head.

**Comment by jhabdas** [comment]
@ForbodingAngel I'm not going to get into it. But the OP is talking about README. Those are static files, Bruh.

**Comment by cirosantilli** [comment]
@jhabdas while discussing if the feature should be implemented or not is perfectly valid, I still want this, and I'd rather keep it open :-)

Also feel free to give a thumbs down to the issue if you think it should not be implemented.

Cheers.

**Comment by jhabdas** [comment]
Downvoted yesterday while offering a path forward to those who might land here and become frustrated this is still open like @AzizSaleh. That said, the real problem here seems more so to do with standardization than of semantics. And the more we press a technology to do things it wasn't designed to do the less interoperable it will likely become.

For example, when I go to StackOverflow these days my fenced code blocks (non-standard) look terrible, forcing me to relearn SO's particular syntactical desires, which seem to be changing.

FWIW Markdown is beautiful because it solves a complex problem in a simple way, which clearly sets the boundaries for what it is and is not capable of (such as producing a TOC) and makes it incredibly fast and interoperable. Take that away and we'll have markdown soup. I already can't use Kramdown block attributes in my other processors, nice as they are, so I'm rather glad Hugo, for example, chose to solve this problem outside the proverbial "kernel".

**Comment by ckpinguin** [comment]
@jhabdas
"And the more we press a technology to do things it wasn't designed to do the less interoperable it will likely become."
I totally agree on that sentence. Anyway the topic is not about changing Markdown, nobody would suggest that for such a function. It's about "a button" (aka a function available on Github) that generates a bit of markdown at the top. I also agree that this is already available through several npm-modules and other programs and I already used [doctoc](https://github.com/thlorenz/doctoc) for that. Still, I see myself editing Markdown directly on Github, and the longer the Files get, the more comfortable such a function would be.

**Comment by jhabdas** [comment]
@ckpinguin Glad you agree. And a button would be nice. How it's implemented should be left up to the implementor. The OP is suggesting a change to markdown here, and this issue is tagged markdown, which is why I jumped in to discourage driving the implementation and closing _this particular_ issue (I saw it was confusing and upsetting people, and ranks quite high in Web Search). My .02 is this issue should be closed or the OP should edit his request.

**Comment by ckpinguin** [comment]
Ups, you're right. To be honest I didn't fully read the OP, I just jumped on the train "I want TOC!" :rofl:

**Comment by dcleao** [comment]
When I upvoted this issue I really wanted a markdown language feature to generate a toc. This feature would be perfectly acceptable for a static language like Markdown. Just see the success of `:toc:` in AsciiDoc. Now, if you don't want  to try to help push the Markdown standard forward, that's ok. Perhaps we just should be "complaining" elsewhere.

**Comment by jhabdas** [comment]
https://productpains.com/product/github/

**Comment by jhabdas** [comment]
https://github.com/dear-github/dear-github

**Comment by jtianling** [comment]
+1
and I add a script to add the toc

https://github.com/jtianling/toc-auto-add

Sorry for the readme writen by Chinese,

**Comment by jhabdas** [comment]
@jtianling that's the way to do it. great job!

**Comment by chaseWillden** [comment]
@tukkek If that's true, that's a shame on GitHub. GitHub should make a stronger voice with CommonMark GitHub is a large consumer of CommonMark's product.

**Comment by jhabdas** [comment]
Influential stakeholder opinion. It can make a product great, and it can lead an implementation in the wrong direction as well. But it sounds like they forked CM, TOC is obviously a product pain, clobbering a file can have, ahem, implications. I wonder if this is a job for something like [StaticMan](https://staticman.net) and Git Hooks if [@jtianling solution](https://github.com/isaacs/github/issues/215#issuecomment-287985587) doesn't fit the bill. That or GH makes the Wiki better, responsive, and auto-generates there like they do with the list of pages. IDK

**Comment by tukkek** [comment]
> If that's true, that's a shame on GitHub

I don't know how it's not. If you follow my link, a GH system administrator literally is telling "this is not our problem anymore, take it somewhere else".

@jhabdas, yes I agree that it must be complicated because otherwise we'd have it already, right? On this issue alone there are over 300 likes towards this feature and probably a dozen or more ad-hoc implementations. If that doesn't how you how huge a need this is for the community, then nothing else will make you see it. This is what GH lives for, providing a platform that solves technical issues, it's what they do.

If GH came out with "this is not a priority because of this and that", "this is so complicated to implement that it's not feasible" or 'wikis are not core to our product", I could respect their position. The only position we were given though is something like "take it elsewhere". This is tremendously bad PR coming from a giant, model website like GitHub

**Comment by jhabdas** [comment]
I hear you. The way I see it is people want TOC. And everyone is rallying behind it. But if the technical feasibility is Nil, it's not going to happen. The pressure caused as a result, bad or good, will lead to innovation. It's a law. And I don't want to get down on GH engineers, look what they've given us, for free. But they're also humans. And sometimes humans struggle to think creatively under duress. Perhaps rather than leasing a specific implementation, someone requests a User Story and frames their needs. I can almost guarantee stepping back from what looks to be the actual problem will help individuals get creative, and ease the pressure.

**Comment by tukkek** [comment]
> the technical feasibility is Nil, it's not going to happen

That's non-sense. Look how many solutions are on this thread already! It's
definitely a far cry from "Nil".

**Comment by jhabdas** [comment]
@tukkek not only did you pick the first thing from my argument you could disagree with to disagree, you failed to respond constructively. Its no wonder this issue is still open. #unsubscribe

**Comment by AzizSaleh** [comment]
I can't believe this thread is still running. I still hold my position that not enough "paid" members are requesting this feature which is why GitHub doesn't care. There is no other explanation, definitely when the actual implementation exists in many languages. If GitHub developers are having a hard time coming with a TOC implementation, I am pretty sure there are thousands (including myself) who will be ready to write it up for them in which language they prefer. Doesn't take a rocket scientist to do this imo.

**Comment by abalter** [comment]
You would need to buy me a lot of gin and tonics to convince me there there are any technical difficulties. There are so many different implementations of TOCs for various markup languages including Common Mark, both back end and front end. Please, if you want to buy me all those gin and tonics let me know. I'll be happy to listen. But if not, I'd like to stop hearing about the technical obstacles. Period.

**Comment by abalter** [comment]
@AzizSaleh -- I AM a paid member!

**Comment by abalter** [comment]
https://github.com/github/markup/issues/904#issuecomment-290143900

**Comment by abalter** [comment]
https://github.com/github/markup/issues/1037

**Comment by AzizSaleh** [comment]
@abalter There are a few like yourself who like this -if you scroll up you can see who they are. Even a few paid members isn't enough imo. I am not a paid member, so I wouldn't know if there are other venues the few of you can make these requests. Hopefully someone that cares will listen.

**Comment by jtianling** [comment]
@AzizSaleh -- I AM a paid member too! After I writen a solution for it, I prefer github give a official solution.

**Comment by tukkek** [comment]
Another official update from GH staff on this issue: "we're handling it internally, not through GitHub issues" https://github.com/github/markup/issues/1036#issuecomment-290266200

**Comment by chaseWillden** [comment]
@tukkek It's not bad to have GitHub issues. If it's being handled internally, it may be an eternity before it's actually implemented.

**Comment by aensley** [comment]
Anyone still waiting for GitHub to do this natively, give this add-on a shot: https://github.com/arthurhammer/github-toc

It's perfect. It just works and works exactly the way I needed it to.

**Comment by bxk-sonavex** [comment]
+1

**Comment by raychenon** [comment]
There are lots of solutions TOC creation script put there.
I created this website https://tableofcontents.herokuapp.com so you don't have to install anything.

It took me 1h to write the initial code. A few days to write the web app. 5 minutes to put on the cloud with a PaaS. Still struggling to push on a bare-metal cloud.

The project just started. You open an issue or share any suggestions on https://github.com/raychenon/auto-toc


> ### Don't find fault, find a remedy. 
> Henry Ford

**Comment by raychenon** [comment]
My online table of contents generator http://tableofcontent.eu/
You don't have to install anything

**Comment by a-i-joe** [comment]
+1

**Comment by rfree-d** [comment]
It's $current_year and people still have to manually create ToC on most popular open-source software hub. Sad!

**Comment by rfree-d** [comment]
@AzizSaleh 

>  I am pretty sure there are thousands (including myself) who will be ready to write it up for them in which language they prefer. Doesn't take a rocket scientist to do this imo.

Maybe there are more pressing issues to handle in the Wiki area, like say making sure developers group is diverse enough.

Option to make Wikipage wider then for 800x600 screen, also is not existing 😞

**Comment by jasonkuhrt** [comment]
This wasn't obvious to me so maybe its not obvious to others but Github supports AsciiDoc which in turn supports great TOC functionality. Its kind of hard to find docs on this so here's an example of what your readme would need to contain:

```
:toc: macro
:toc-title:
:toclevels: 9

# Title

## A

### A2

## B

### B2

```

Here is an example in one of my own repos: https://github.com/littlebits/react-popover/blob/master/README.adoc

There are lots of nice little things about AsciiDoc but only its good TOC functionality for me typically warrants the change to another markup language. I'd love to be able to stay with markdown and win from the great tool/editor support etc.. For now though AsciiDoc has been working well for me.

**Comment by holvonixAdvay** [comment]
@jasonkuhrt  - Frankly 'use AsciiDoc' was the best advice I've seen here - thanks.  So much more powerful than markdown, and not sure if anyone realizes GitHub supports it...

**Comment by jasonkuhrt** [comment]
@holvonixAdvay Super happy to have helped!! 👍

**Comment by pedro-mass** [comment]
@jasonkuhrt - is there anything special you need to do to get that toc macro to work? Can it be used with markdown in other parts of the page?

**Comment by jasonkuhrt** [comment]
@pedro-mass Its an AsciiDoc only feature. You can place the TOC macro wherever you want. I think it will only show titles that appear below where you use the macro.

**Comment by danielehrlich** [comment]
+1 If GitHub made this is would make the world such a better place. There would be better documentation, and therefore less screw-ups and just generally happier people.

**Comment by brazilbean** [comment]
We've been wanting to move our software documentation out of Confluence and into GitHub. The biggest gain to be had is that the documentation will be versioned with the source (+++1). We're wary though that markdown lacks all of basic essentials, such as TOC support. (-1). 

In the end, we'll still give markdown a try, but having TOC support would sure be nice. :)

**Comment by genio** [comment]
I hate to pile on here, GitHub. But, it's now 2018. Can we _please_ have a way to auto-generate TOCs?

**Comment by mckaygerhard** [comment]
github its a crap, take a look to a real open source model development, the TOC generation on the gambas wiki logic are great, event generated from the title words, are using tags based on order

at leas gitlab have more enphasis..

**Comment by gtaifu** [comment]
After looking around for the tool to generate the gfm TOC, I found the tool [github-markdown-toc]((https://github.com/ekalinin/github-markdown-toc) by @ekalinin . Though the code is simply, it runs in Bash originally targets Linux.

Another tool is [markdown-toc](https://github.com/jonschlinkert/markdown-toc) by @jonschlinkert, but it   depends on [remarkable](https://github.com/jonschlinkert/remarkable). 

I would like to have a snippet running locally also at Windows. So, I created a small project [gfm_toc_generator](https://github.com/gtaifu/gfm_toc_generator) based on Python3. (I wish I did not recreated the wheel.) 

I will make it importable from Python and also a Sublime package later. Hope this would be helpful to you.

Hope this would be helpful to you.

**Comment by mckaygerhard** [comment]
the project from @ekalinin are better, python3 its a messe up of complicated dependences that changes on every api upgrade.. also i do not use a poor level operatin system like guindows

**Comment by angelcervera** [comment]
I don't like any of the listed solutions. As @cirosantilli said, It must be a feature included in GitHub markdown. **It is a basic feature to write readable documentation!** The rest of options are not solutions, are workarounds.

**Comment by mckaygerhard** [comment]
all are workaround due some want ruindows-like solutions compatible, and well @angelcervera github are deployed in unix-like, and almost all wiki-based software are develop in a unix-like in mind..

a web-based solution are not provided wuickly due this main reason, the second are due github has a terrible error, the anchors are based on the titles, and not on the marks.. like i suggested was made for the gambas-wiki software a good made software (of course only works well on unix-like)

**Comment by eeholmes** [comment]
I just wanted an autogenerated toc on my ReadMe file on my repository, so I used @jasonkuhrt suggestion to use AsciiDoc.

1. I renamed my README file from README.md to README.adoc
2. I added this to the top of my README file
```
:toc: macro
:toc-title:
:toclevels: 99

# Table of Contents
toc::[]
```

30 seconds and presto -- TOC at the top of my README.  My readme file is simple.  Headings, code.  I didn't do anything but rename md file and add those toc lines to top of readme file.

**Comment by jasonkuhrt** [comment]
If you are using Atom editor and if Asciidoc is undesirable (team setting, have to learn new things, not in standards, etc.) I have found that using an editor plugin can help a lot. I'm using https://github.com/t9md/atom-markdown-toc-auto.

One downside is that collaborators need to be synced with you on the editor/plugin choice.

**Comment by dbkaplun** [comment]
I wrote an extension to do this for Chrome a while ago: [dbkaplun/github-markdown-outline-extension](https://github.com/dbkaplun/github-markdown-outline-extension)

**Comment by AndyWendt** [comment]
I first upvoted this over three years ago...  I'm getting the vibe that this isn't a high priority haha

**Comment by Jofairden** [comment]
It would be really nice to have `__TOC__` support for Markdown

**Comment by ArtSabintsev** [comment]
Just adding my +1

**Comment by rfree-d** [comment]
Microsoft 👏 make 👏 github 👏 staff 👏 code 👏 this 👏 shit 👏

**Comment by Gnappuraz** [comment]
+1

**Comment by davidrdark** [comment]
+1

**Comment by kalbasit** [comment]
Please stop adding your +1 comments!! @isaacs can you please lock this conversation? Thank you!

**Comment by jujumo** [comment]
Anyway, I moved to Asciidoc, because github implements the TOC on it. I do not regret it,

**Comment by textractor** [comment]
Is Macrosoft reading this? Please just change all of this to Office so we can all go home.

**Comment by YingmingHu** [comment]
+1

**Comment by rfree-d** [comment]
4 years later...

![4years](https://user-images.githubusercontent.com/18722764/44949017-302dcc00-ae29-11e8-91c3-b61942fd7e2e.jpg)

**Comment by mckaygerhard** [comment]
in any case! now i use only gitlab!

**Comment by Jucesr** [comment]
> You can automatically generate TOC via http://doctoc.herokuapp.com/. There is also an npm package available [here](https://github.com/thlorenz/doctoc/blob/master/README.md) to do it on the command line.
> 
> But yes, these are ultimately just manual methods because you need to manually copy and paste the TOC into the markdown. But one benefit of it is that the toc is maintained everwhere, even across apps that don't use GFM for rendering.

I am afraid I just broke it

**Comment by peterpeterparker** [comment]
https://magnetikonline.github.io/markdown-toc-generate/ from @magnetikonline is a cool tool to generate table of contents

**Comment by topdeclassified** [comment]
I agree this should be a BASIC feature. I can do this in Microsoft Visual Code why not here?

**Comment by CryfryDoesGaming** [comment]
MediaWiki style has an option for making Table of Contents.

**Comment by tgoodlet** [comment]
For those interested restructured text (`.rst`) [has this built-in](http://docutils.sourceforge.net/docs/ref/rst/directives.html#table-of-contents) FYI.

**Comment by sompylasar** [comment]
While in the Wiki Edit (`/_edit`) Preview mode, run this snippet in the browser console to get a Markdown of a Table of contents:
```js
console.log(
  '\n\n\n> **Table of contents**\n> \n' +
  Array.from(document.querySelectorAll('h1 > a, h2 > a, h3 > a')).map((a) => (
    {'H1':'> * ','H2':'>   * ','H3':'>     - '}[a.parentNode.tagName] +
    `[${a.parentNode.innerText.trim()}](${a.hash})`
  )).join('\n') +
  '\n\n\n'
);
```

Example output from a page with H1 (`#`), H2 (`##`), H3 (`###`):
```

> **Table of contents**
> 
> * [Test](#test)
>   * [Partial data dump](#partial-data-dump)
>     - [Technical details](#technical-details)
>   * [Full data dump](#full-data-dump)
>   * [Sanity checks](#sanity-checks)

```

Which is rendered as:

> **Table of contents**
> 
> * [Test](#test)
>   * [Partial data dump](#partial-data-dump)
>     - [Technical details](#technical-details)
>   * [Full data dump](#full-data-dump)
>   * [Sanity checks](#sanity-checks)

**Comment by imthenachoman** [comment]
I got frustrated at the lack of a quick and easy online solution so I created this: [https://imthenachoman.github.io/nGitHubTOC](https://imthenachoman.github.io/nGitHubTOC/).

**Comment by NBTX** [comment]
Two things:
1. Perhaps it might be worth moving this discussion to the [GitHub Community Forums](https://github.community)
2. Failing any input or feedback from GitHub, what if we created an online service that you could just drop your repository links or webhooks into and it would automatically generate a Table of Contents in your README and update it whenever a push is made?

**Comment by ildar-shaimordanov** [comment]
In addition to the existing bunch of tools converting Markdown documents to the TOC, I implemented one more tool in Perl: https://github.com/ildar-shaimordanov/git-markdown-toc.

**Comment by dreftymac** [comment]
# Quick Summary (TL;DR)

* automatic TOC is not (yet) a feature in GH flavored markdown
* here is a summary of what people have said about this

## Alternate approaches

* asciidoc ;; https://asciidoctor.org/docs/what-is-asciidoc/
* various one-off cut-and-paste websites

## Editor-based solutions

* vscode      ;; https://marketplace.visualstudio.com/items?itemName=AlanWalk.markdown-toc
* vim         ;; https://github.com/mzlogin/vim-markdown-toc
* sublimetext ;; https://packagecontrol.io/packages/MarkdownTOC
* atom        ;; https://atom.io/packages/markdown-toc

**Comment by svoop** [comment]
@dreftymac Maybe worth mentioning that the GitHub wiki is powered by Gollum which features `[[_TOC_]]`, so not suppressing this feature seems to be a solution, too.

**Comment by joric** [comment]
I need this desperately (a built in feature, not the 3rd-party stuff, or at least a Chrome extension), so bump.

**Comment by kdeldycke** [comment]
My personal take on generating and maintaining a TOC in Markdown consist in using a GitHub Action: https://github.com/kdeldycke/awesome-iam/blob/e21d96985b4cb64b8eddacd16819b87aed2ac632/.github/workflows/lint.yaml#L11-L31 .

**Comment by mrts** [comment]
[Markdown Preview Enhanced](https://shd101wyy.github.io/markdown-preview-enhanced/#/toc) works beautifully for this. I'm not sure GitHub should do this - maybe IDE support is enough?

**Comment by svoop** [comment]
@mrts TOC is a feature of Gollum (which powers the wiki) and has to be explicitly disabled. It should be 1 LOC and a bit of styling to enable TOC, so no, definitely better to do this server-side and not force authors to use IDEs.

**Comment by TradeIdeasPhilip** [comment]
> You can automatically generate TOC via http://doctoc.herokuapp.com/. There is also an npm package available [here](https://github.com/thlorenz/doctoc/blob/master/README.md) to do it on the command line.
> 
It crashed when I tried it.

Welcome to 2020.  Everything is free.  But any time you want something you have to try 10 different websites to find the one that actually works.

**Comment by filiptronicek** [comment]
I found this today, it looks like it is a solution to this problem:
![image](https://user-images.githubusercontent.com/29888641/112554719-18dbd300-8dc7-11eb-92a9-5840590fc1c2.png)

It's just beside `README.md` in every GH repo!

**Comment by svoop** [comment]
@filiptronicek Nice, now all that's left is to add a similar button to other places where Markdown is rendered, most notably the wikis. Given this took 7 years in the making, I'm not gonna hold my breath. 😄

**Comment by SiddharthShyniben** [comment]
> I found this today, it looks like it is a solution to this problem:
> ![image](https://user-images.githubusercontent.com/29888641/112554719-18dbd300-8dc7-11eb-92a9-5840590fc1c2.png)
> 
> It's just beside `README.md` in every GH repo!

But I don't see it in _every_ repo. Who knows?

**Comment by filiptronicek** [comment]
> 
> 
> > I found this today, it looks like it is a solution to this problem:
> > ![image](https://user-images.githubusercontent.com/29888641/112554719-18dbd300-8dc7-11eb-92a9-5840590fc1c2.png)
> > It's just beside `README.md` in every GH repo!
> 
> But I don't see it in _every_ repo. Who knows?

I noticed that too, I'm currently experimenting, but an example of a repo that has a README that is long and should be IMO indexed is [steveseguin/obsninja](https://github.com/steveseguin/obsninja) (stars also shouldn't be a problem, this one has nearly a thousand).

**Comment by botatoes** [comment]
Hi all! PM at GitHub here that on this feature. If you didn't see it at the time in previous comments, you should be able to see it now. We rolled it out 100% as of a few hours ago. 

Just letting you all know we will be following up with adding support for this to other places as well. Currently there's plans for: 
- non-README rendered pages
- Wikis
- possibly Markdown Previews for rendered Markdown Pages.

Let me know if you have any feedback.

**Comment by xantari** [comment]
@botatoes When will the wiki's support be added?

**Comment by TPS** [comment]
@botatoes Thanks also for implementing this for the mobile _site_.  Any word when it'll be on the mobile _apps_?

**Comment by JeremyCaney** [comment]
@botatoes: Thank you! This is a nicely considered implementation. I especially appreciate that the menu is sticky so that it can easily be referenced even after you've scrolled down the page—and that it highlights what section you're currently viewing as you scroll down the page. I was just hoping for a tag that would embed a table of contents within the markdown, but this level of deep integration with the product is a much better approach. I look forward to pushing some commits to my repos as I rip out previously hard-coded tables of contents.

**Comment by albertkun** [comment]
@botatoes this is such a great feature!!! I am using GitHub to teach a class and am very happy about the roll out of the TOC! One request, is it possible to have an option in the repo for it to be "expanded on default" and have it on the right side so it doesn't cover the title pages?

something like this would be ideal:
![image](https://user-images.githubusercontent.com/8574425/113200224-e72a9800-921c-11eb-8919-6b175607214d.png)

regardless, thank you for this amazing feature!!

**Comment by captn3m0** [comment]
Minor concerns with the new feature:

1. The new feature only works on the website, not on the Mobile App (I tested on iOS, might be supported on Android).
2. It also feels very non-obvious (unless I was looking for it, I'd never have clicked it). Especially for projects on GitHub that don't have developers (regular GitHub users) as the primary audience (stuff like https://github.com/adamdriscoll/awesome-covid19-resources for eg).

**Comment by albertkun** [comment]
> 
> 
> @botatoes Thanks also for implementing this for the mobile _site_. Any word when it'll be on the mobile _apps_?

It works great on Firefox Android web browser.

**Comment by SiddharthShyniben** [comment]
@captn3m0 It doesn't appear on Android too

**Comment by botatoes** [comment]
@captn3m0 Our web implementations is the most to date. The mobile app team has their own backlog which means features aren't always aligned to ship at the same time. TOC should work on web mobile view though. Let me know if that's not the case for you.

**Comment by thorade** [comment]
Seems this feature was announced here:
https://twitter.com/github/status/1376636651963842562
Is there any documentation elsewhere?

**Comment by botatoes** [comment]
@thorade there will be a changelog and updated docs. I'm just waiting for the additional non-README Markdown support coming either late this week or early next week before I push out the docs.

**Comment by botatoes** [comment]
Here's the link to the changelog https://github.blog/changelog/2021-04-13-table-of-contents-support-in-markdown-files/ and docs update will come soon. This is now supported for all Markdown files (in the browser).

**Comment by TPS** [comment]
@botatoes Thanks very much!  Is there an ETA for TOC support in the GitHub mobile _apps_?

**Comment by CatalanCabbage** [comment]
[[_TOC_]] *

# No, this shouldn't be closed!  
True, [this](https://github.com/isaacs/github/issues/215#issuecomment-808595170) is a nice feature.  
What we need is [something like what GitLab supports](https://docs.gitlab.com/ee/user/markdown.html#table-of-contents) - no-nonsense, just a Table of Contents on top.  

## Why?
Not obvious AT ALL, and does not seem like a part of the Markdown content.  
A Table of Contents needs to be a part of the Markdown content, since it refers to the sections in the comment itself.  

_Image for reference:_
<img src="https://user-images.githubusercontent.com/45961072/115148163-fa8e8d80-a07b-11eb-970c-0f53ba54e7c1.png" height="399px" width="624px"></img>
There's NO WAY I could ever imagine that clicking that takes me to the ToC of the MD contents.  
I first thought it would show a list of files on the same level for file navigation, and I've been using GitX for a long time.

---
## But...
Why complicate this? I mean, it's cool and all, but why not just go with [this](https://docs.gitlab.com/ee/user/markdown.html#table-of-contents), the
- the simple, 
- more effective, 
- flexible (I can choose not to give a ToC), 
- more obvious (if it's on top and a part of the MD file, it's the ToC of the MD file, duh), 
- more easy to use by _literally_ all people (that's the global format, intuitive)  
  (and by all people I mean - even if Robinson Crusoe came back from 30 years on an island and saw this, he'd get it)
way?

---
## Suggestion:
If you want to make the traditional ToC better, I would suggest that we implement something like Wikipedia's ToC with a hide/show option.  
We can choose to give a ToC using a `[[_TOC_]]` or whatever, and then you can render it. Plus we can keep the cool button if you want. 😄 

<img src="https://user-images.githubusercontent.com/45961072/115148519-98368c80-a07d-11eb-914d-e7ef8d41bf49.png" height="700px" width="624px"></img>
[reference](https://en.wikipedia.org/wiki/Cache_replacement_policies)

---

* <sup>I have a dream! That one day when ToC is implemented better, this will be rendered as a proper ToC</sup>

**Comment by xantari** [comment]
Still not seeing table of contents on Github Wiki pages :(

**Comment by svoop** [comment]
@xantari Don't worry, it'll be there soon enough, give it another – say – 8 years. :wink:

**Comment by botatoes** [comment]
@CatalanCabbage The reason why we've picked the UI route first instead of injecting Markdown is because GitHub README.md files are rendered elsewhere outside of GitHub UI and injecting Markdown might create unexpected behaviors. That being said, we'll still be considering it for the future as a part of our improve GitHub Flavored Markdown effort. 

@xantari I'm driving proposals for Wiki improvements right now and this is definitely on the list but no timeline yet 😅

**Comment by CatalanCabbage** [comment]
@botatoes if that's the reason, then is it possible to 
- Make it really noticeable so that it's intuitive for non-devs
- Allow authors to enable or disable it somehow

The problems with it now are that it's 
- Hard to find
- Hides content when it's open, so you cannot read the file while keeping this open

Something like this (either on the left or right), prominent like you see in Docs when it's open?   

Example from Jest:  
![image](https://user-images.githubusercontent.com/45961072/115306926-25650880-a186-11eb-963c-4a31d42f777b.png)

**Comment by hectoralicea** [comment]
I just got this working on GitLab via `[[_TOC_]]` tag.  Works great in my editor as well (using Typora).  Very surprised GitHub does not support this simple feature.

**Comment by xantari** [comment]
@hectoralicea That’s also how Azure DevOps does it as well it’s so easy

**Comment by mxmilkiib** [comment]
Wiki ToC issue is #504
